### PR TITLE
Backport: Duotone changes

### PR DESCRIPTION
**Pull request by MaggieCabrera**
This PR backports the changes made to the duotone related PHP files. I grouped them together instead of having one PR for each change because they would all depend on each other, which made more sense. This PR depends on the stabilized selectors API to be backported for the changes to work.

Trac ticket: https://core.trac.wordpress.org/ticket/58555
Backports:
- [x] https://github.com/WordPress/gutenberg/pull/49103
- [x] https://github.com/WordPress/gutenberg/pull/49232
- [x] https://github.com/WordPress/gutenberg/pull/49212
- [x] https://github.com/WordPress/gutenberg/pull/49239
- [x] https://github.com/WordPress/gutenberg/pull/49393 I only backported the changes to `class-wp-duotone-gutenberg.php` because the rest seemed dependent on other unrelated changes
- [x] https://github.com/WordPress/gutenberg/pull/49423 I only backported the changes to `class-wp-duotone-gutenberg.php` because the rest seemed dependent on other unrelated changes
- [x] https://github.com/WordPress/gutenberg/pull/49700
- [x] https://github.com/WordPress/gutenberg/pull/49702
- [x] https://github.com/WordPress/gutenberg/pull/49705
- [x] https://github.com/WordPress/gutenberg/pull/49706
- [x] https://github.com/WordPress/gutenberg/pull/50714
- [x] https://github.com/WordPress/gutenberg/pull/51397

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/block-supports/duotone.php
Comment (MaggieCabrera) [inline tests/phpunit/tests/block-supports/duotone.php]: I don't know if this is the correct place for the test. We didn't have the old test that was removed in https://github.com/WordPress/gutenberg/pull/49103 and this is where it made sense to me...

**Comment by ajlende** [comment]
[Some of the duotone code](https://github.com/MaggieCabrera/wordpress-develop/blob/9a285578bfbcd352a5fd3c6c64fad0d419d93e8a/src/wp-includes/class-wp-duotone.php#L768-L802) depends on https://github.com/WordPress/gutenberg/pull/46496 getting backported first.

**Comment by tellthemachines** [comment]
@MaggieCabrera is there a Trac ticket for these changes? If not, we'll need to create one and add it in the PR description.

**Comment by MaggieCabrera** [comment]
> @MaggieCabrera is there a Trac ticket for these changes? If not, we'll need to create one and add it in the PR description.

No, I was planning on doing that today

File: src/wp-includes/block-supports/duotone.php
Comment (audrasjb) [inline src/wp-includes/block-supports/duotone.php]: This line should be removed as it's a duplicate.

File: src/wp-includes/class-wp-duotone.php
Comment (audrasjb) [inline src/wp-includes/class-wp-duotone.php]: As per WP Documentation Standards, should be "Takes", not "Take" (third-person verb)
https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#language

File: src/wp-includes/class-wp-duotone.php
Comment (audrasjb) [inline src/wp-includes/class-wp-duotone.php]: As per WP Documentation Standards, should be "Checks", not "Check" (third-person verb)
https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#language

File: src/wp-includes/class-wp-duotone.php
Comment (audrasjb) [inline src/wp-includes/class-wp-duotone.php]: As per WP Documentation Standards, should be "Gets", not "Get" (third-person verb)
https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#language

File: src/wp-includes/class-wp-duotone.php
Comment (audrasjb) [inline src/wp-includes/class-wp-duotone.php:747]: Was it really included AND deprecated in 6.3? 🤔

File: tests/phpunit/tests/block-supports/duotone.php
Comment (audrasjb) [inline tests/phpunit/tests/block-supports/duotone.php:25]: We probably need a Docblock here :)

File: tests/phpunit/tests/block-supports/duotone.php
Comment (audrasjb) [inline tests/phpunit/tests/block-supports/duotone.php:42]: We probably need a Docblock here :)

**Review by audrasjb — CHANGES_REQUESTED** [review]
Thanks for the PR!
I left some comments concerning this changeset :)

File: src/wp-includes/class-wp-duotone.php
Comment (MaggieCabrera) [reply src/wp-includes/class-wp-duotone.php:747]: @ajlende can you have a look?

File: tests/phpunit/tests/block-supports/duotone.php
Comment (MaggieCabrera) [reply tests/phpunit/tests/block-supports/duotone.php:25]: Let me know if this is enough, thank you!

File: src/wp-includes/class-wp-duotone.php
Comment (ajlende) [reply src/wp-includes/class-wp-duotone.php:747]: This is accurate. The comment explains why.

> Exported for the deprecated function wp_get_duotone_filter_id().

In the process of refactoring duotone into a class in this last release cycle, `wp_get_duotone_filter_id` was deprecated. Pieces of the implementation were moved into private methods, so in order to keep the deprecated functions available they were added as deprecated public methods for the class.

File: tests/phpunit/tests/block-supports/duotone.php
Comment (ramonjd) [inline tests/phpunit/tests/block-supports/duotone.php:20]:
```suggestion
	 * @ticket 58555
	 *
```

I think we could also add `@ticket 58555`, referencing the [ticket](https://core.trac.wordpress.org/ticket/58555) of this PR

File: tests/phpunit/tests/block-supports/duotone.php
Comment (ramonjd) [inline tests/phpunit/tests/block-supports/duotone.php:37]: `@ticket 58555` as above

**Comment by ramonjd** [comment]
> [Some of the duotone code](https://github.com/MaggieCabrera/wordpress-develop/blob/9a285578bfbcd352a5fd3c6c64fad0d419d93e8a/src/wp-includes/class-wp-duotone.php#L768-L802) depends on https://github.com/WordPress/gutenberg/pull/46496 getting backported first.

Backport PR that includes https://github.com/WordPress/wordpress-develop/pull/4619 over here: 
- https://github.com/WordPress/wordpress-develop/pull/4655

It also includes changes from https://github.com/WordPress/gutenberg/pull/49427

File: src/wp-includes/class-wp-duotone.php
Comment (audrasjb) [reply src/wp-includes/class-wp-duotone.php:747]: Ok thanks! I understand the idea.

**Comment by ramonjd** [comment]
Would this PR also need backporting for duotone to work?

- https://github.com/WordPress/gutenberg/pull/50678

cc @ajlende

File: tests/phpunit/tests/block-supports/duotone.php
Comment (peterwilsoncc) [inline tests/phpunit/tests/block-supports/duotone.php]:
```suggestion
class Tests_Block_Supports_DuoTones extends WP_UnitTestCase {
```

No WP prefix in test suite.

File: src/wp-includes/block-supports/duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/block-supports/duotone.php:42]: This block ought to be in the default-filters file.

File: src/wp-includes/block-supports/duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/block-supports/duotone.php]: Deprecated functions ought to be moved to src/wp-includes/deprecated.php

File: src/wp-includes/class-wp-duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-duotone.php]:
```suggestion
	 * Clamps a value between an upper and lower bound.
```

File: src/wp-includes/class-wp-duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-duotone.php]:
```suggestion
	 * @link https://github.com/omgovich/colord/blob/3f859e03b0ca622eb15480f611371a0f15c9427f/src/helpers.ts#L23 Sourced from colord.
```

File: src/wp-includes/class-wp-duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-duotone.php]:
```suggestion
	 * @param float $number The number to clamp.
	 * @param float $min    The minimum value.
	 * @param float $max    The maximum value.
```

File: src/wp-includes/class-wp-duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-duotone.php:174]:
```suggestion
	 * Processes and clamps a degree (angle) value properly.
```

This applies throughout -- the docblocks are used to generate developer.wordpress.org so need to explain what the method does.

File: src/wp-includes/class-wp-duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-duotone.php]:
```suggestion
	 * @link https://github.com/omgovich/colord/blob/3f859e03b0ca622eb15480f611371a0f15c9427f/src/helpers.ts#L32 Sourced from colord.
```

Again, this applies throughout -- a link annotation is better for an external reference.

File: src/wp-includes/class-wp-duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-duotone.php]:
```suggestion
	 * Takes the inline CSS duotone variable from a block and return the slug.
	 *
	 * Handles styles slugs like:
	 * var:preset|duotone|blue-orange
	 * var(--wp--preset--duotone--blue-orange)
	 *
```

File: src/wp-includes/class-wp-duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-duotone.php]:
```suggestion
	 * Enqueue custom filter assets for the page.
	 *
	 * Includes an SVG filter and block CSS declaration.
```

File: src/wp-includes/class-wp-duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-duotone.php]:
```suggestion
	 * Enqueue preset assets for the page.
	 *
	 * Includes a CSS custom property, SVG filter, and block CSS declaration.
```

File: src/wp-includes/class-wp-duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-duotone.php]:
```
/*
 * WP Coding standards are to use  this 
 * format for multiline comments
 */
```

:1234: Applies throughout

File: src/wp-includes/class-wp-duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-duotone.php]:
```suggestion
	 * Get all possible duotone presets from global and theme styles.
	 *
	 * Store each item as slug => [ colors array ]
	 * We only want to process this one time. On block render we'll access and output only the needed presets for that page.
```

File: src/wp-includes/class-wp-duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-duotone.php]:
```suggestion
	 * Adds the duotone SVGs and CSS custom properties to the editor settings.
	 *
	 * This allows the properties to be
	 * pulled in by the EditorStyles component in JS and rendered in
	 * the post editor.
```

But formatted nicely.

File: src/wp-includes/class-wp-duotone.php
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-duotone.php]: Again this needs to be split in to a summary and longer description/

When the developer docs are generated the first line is a short summary displayed below the title, followed by a longer description that goes in to detail. See https://developer.wordpress.org/reference/classes/wp_query/get_posts/ fpr example.

:1234: This chnage is needed in numerous places so I'll get you to reeveluate them rather than continue adding noise to the PR

File: tests/phpunit/tests/block-supports/duotone.php
Comment (peterwilsoncc) [inline tests/phpunit/tests/block-supports/duotone.php:86]: Data provide should go below the test and have the docblock

```
/**
 * Data provider.
 *
 * @return array[].
 */
```

:1234: Applies below too

**Review by peterwilsoncc — COMMENTED** [review]
I've added a few notes relating to the docblocks and locations.

I've used :1234: to indicate that the item applies in multiple locations.

File: src/wp-includes/class-wp-duotone.php
Comment (peterwilsoncc) [reply src/wp-includes/class-wp-duotone.php:747]: Is it worth adding an `@internal` annotation to the docblocks for items introduced and deprecated in the same version like this?

That will prevent the docs from been added to developer.wordpress.org.

**Comment by MaggieCabrera** [comment]
> Would this PR also need backporting for duotone to work?
> 
>     * [Fix custom duotone filters in frontend gutenberg#50678](https://github.com/WordPress/gutenberg/pull/50678)
> 
> 
> cc @ajlende

Apparently that was already ported to core [here](https://github.com/WordPress/wordpress-develop/pull/4108)

File: src/wp-includes/class-wp-duotone.php
Comment (MaggieCabrera) [reply src/wp-includes/class-wp-duotone.php]: @ajlende can you help with this one?

File: src/wp-includes/class-wp-duotone.php
Comment (MaggieCabrera) [reply src/wp-includes/class-wp-duotone.php]: Thank you, I think I covered them all!

File: src/wp-includes/class-wp-duotone.php
Comment (MaggieCabrera) [reply src/wp-includes/class-wp-duotone.php]: Done, I think I covered all of them

File: src/wp-includes/class-wp-duotone.php
Comment (ajlende) [reply src/wp-includes/class-wp-duotone.php:174]: These methods are all private. Is there a way to not generate documentation for this on developer.wordpress.org?

Edit: Just saw your [other comment](https://github.com/WordPress/wordpress-develop/pull/4619#discussion_r1241375630) about adding `@internal`. Updated in [7a43050](https://github.com/WordPress/wordpress-develop/pull/4619/commits/7a430500005a9b3bd0963ab7a63db57c62821575).

File: src/wp-includes/block-supports/duotone.php
Comment (ajlende) [reply src/wp-includes/block-supports/duotone.php:42]: All of the other block supports (layout and position) have their hooks co-located with the support registration in block-supports. Should those be moved too? Or is it okay having these here?

**Comment by ajlende** [comment]
@peterwilsoncc All of the public methods are to be used exclusively with the hooks that we have defined. The other block supports have `@access private` added to their hooks—can that be used for public methods on the `WP_Duotone` class or should the whole class be marked private somehow?

File: src/wp-includes/class-wp-duotone.php
Comment (ajlende) [reply src/wp-includes/class-wp-duotone.php]: Updated in [0b06740](https://github.com/WordPress/wordpress-develop/pull/4619/commits/0b067400ce2dd3c95cc955315469ff973cb61b4a).

File: src/wp-includes/class-wp-duotone.php
Comment (ramonjd) [inline src/wp-includes/class-wp-duotone.php:628]: Do all these new methods require

`@since 6.3.0`

?

File: src/wp-includes/class-wp-theme-json.php
Comment (ramonjd) [inline src/wp-includes/class-wp-theme-json.php:147]: Suggested annotation for docblock

`* @since 6.3.0 Replaced value_func for duoone with `null`. Custom properties are handled by class-wp-duotone.php.`

**Review by ramonjd — COMMENTED** [review]
I've tested this manually on top of:

- https://github.com/WordPress/wordpress-develop/pull/4655

Functionality-wise things are working well for me. 

I think it'll just need a final comb from a core committer?

**Comment by ramonjd** [comment]
Hmm some of the failing tests are real though. 

I'm seeing `-wp--preset--duotone--custom-duotone: url('#wp-duotone-custom-duotone');` not being output in `Tests_Theme_wpThemeJson::test_get_stylesheet`

🤷 

I guess we can rerun and debug when and if https://github.com/WordPress/wordpress-develop/pull/4655 is committed and this PR rebased (?)

**Comment by MaggieCabrera** [comment]
I just rebased this after https://github.com/WordPress/wordpress-develop/pull/4655 got merged

**Comment by getsource** [comment]
This may be something you're already checking into, but in case it helps -- it looks like a few of the test failures may be PHPUnit tests that need to be updated:
https://github.com/WordPress/wordpress-develop/actions/runs/5387774760/jobs/9779547969#step:17:328

File: src/wp-includes/class-wp-theme-json.php
Comment (ajlende) [reply src/wp-includes/class-wp-theme-json.php:147]: Fixed in [3265f93](https://github.com/WordPress/wordpress-develop/pull/4619/commits/3265f93001b09a3e88385e768a3e5ccf1e090cc1)

File: src/wp-includes/class-wp-duotone.php
Comment (ajlende) [reply src/wp-includes/class-wp-duotone.php:628]: Fixed in [817b916](https://github.com/WordPress/wordpress-develop/pull/4619/commits/817b916492e3222f0752abe267f6aebc9465ebef)

**Comment by ajlende** [comment]
Alright, I think the remaining failing tests are unrelated to these changes, and all the feedback has been addressed. I've also tested it now that it's been rebased with the selectors API changes. So this should be ready to go except for confirming [one thing](https://github.com/WordPress/wordpress-develop/pull/4619#issuecomment-1608024130):

> The other block supports have `@access private` added to their hooks—can that be used for public methods on the `WP_Duotone` class or should the whole class be marked private somehow?

I added `@access private` to the whole class. I hope that keeps docs from being generated for developer.wordpress.org. The documentation on [docblock tags](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#phpdoc-tags) isn't clear about what prevents docs from being generated.

File: src/wp-includes/block-supports/duotone.php
Comment (peterwilsoncc) [reply src/wp-includes/block-supports/duotone.php:42]: @ajlende let's leave it as is for now (including in this PR). They probably all should be in default-filters but probably a discussion for another time.

**Comment by peterwilsoncc** [comment]
> I added `@access private` to the whole class. I hope that keeps docs from being generated for developer.wordpress.org. The documentation on [docblock tags](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#phpdoc-tags) isn't clear about what prevents docs from being generated.

@ajlende Items with `@access private` have posts generated on the developer docs, see [`_cleanup_image_add_caption()`](https://developer.wordpress.org/reference/functions/_cleanup_image_add_caption/) as an example (chosen at random).

Sorry, I'm unclear of the context for this https://github.com/WordPress/wordpress-develop/pull/4619#issuecomment-1608024130, would you be able to clarify?

**Comment by ramonjd** [comment]
I've rebased this onto trunk (locally) and tested in editor again. 

Looking good functionality-wise! Duotone being output as expected on the frontend.

Thanks for all the hard work on this. 

Is the @private annotation question the only thing holding this up code-wise @peterwilsoncc ?

**Comment by ajlende** [comment]
>> All of the public methods are to be used exclusively with the hooks that we have defined. The other block supports have @access private added to their hooks—can that be used for public methods on the WP_Duotone class or should the whole class be marked private somehow?
>
> Sorry, I'm unclear of the context for this https://github.com/WordPress/wordpress-develop/pull/4619#issuecomment-1608024130, would you be able to clarify?

@peterwilsoncc The exposed public methods should only be called once in a very particular order defined in [block-supports/duotone.php](https://github.com/WordPress/wordpress-develop/pull/4619/files#diff-c58ecce72a1b4165b86fd3af353360ae43a6d4771215f80532e240e224c0a949R43-R63)—they shouldn't really be called otherwise.

This is one of the hooks for layout that has `@access private` that I was talking about—it should only get called during the `render_block_core/image` filter which happens [later](https://github.com/WordPress/wordpress-develop/blob/80e424ad55e0ef4d48959a9c91b226eae45b6a77/src/wp-includes/block-supports/layout.php#L861) in the same file. https://github.com/WordPress/wordpress-develop/blob/80e424ad55e0ef4d48959a9c91b226eae45b6a77/src/wp-includes/block-supports/layout.php#L802-L813

**Review by tellthemachines — APPROVED** [review]
LGTM! Duotone filters working well in editor and front end.

**Comment by tellthemachines** [comment]
committed in [r56101](https://core.trac.wordpress.org/changeset/56101) / [d966798](https://github.com/WordPress/wordpress-develop/commit/d966798bfcf4481053d6206bba787b07e5030775).

**Pull request closed without merging**
